File: app/src/main/java/de/baumann/browser/view/sortlistpreference/MultiSelectDragListPreferenceDialog.java
package de.baumann.browser.view.sortlistpreference;

import android.os.Bundle;
import android.util.SparseBooleanArray;
import android.view.View;
import android.widget.AdapterView;
import android.widget.ArrayAdapter;
import android.widget.ListView;

import androidx.preference.PreferenceDialogFragmentCompat;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

import de.baumann.browser.Ninja.R;

public class MultiSelectDragListPreferenceDialog extends PreferenceDialogFragmentCompat {
    private final MultiSelectDragListPreference preference;

    private DragSortListView listView;

    public MultiSelectDragListPreferenceDialog(MultiSelectDragListPreference preference) {
        this.preference = preference;

        final Bundle b = new Bundle();
        b.putString(ARG_KEY, preference.getKey());
        setArguments(b);
    }


    private boolean[] getSelectedItems() {
        final CharSequence[] entries = preference.entryValues;
        final int entryCount = entries.length;
        final List<String> values = preference.getValues();
        boolean[] result = new boolean[entryCount];

        for (int i = 0; i < entryCount; i++) {
            result[i] = values.contains(entries[i].toString());
        }

        return result;
    }

    private ArrayAdapter<String> adapter;
    private boolean mPreferenceChanged;
    private DragSortListView.DropListener onDrop = new DragSortListView.DropListener() {
        @Override
        public void drop(int from, int to) {
            if (from != to) {
                String item = adapter.getItem(from);
                adapter.remove(item);
                adapter.insert(item, to);
                listView.moveCheckState(from, to);
                mPreferenceChanged = true;
                refreshNewValues();
            }
        }
    };

    private void refreshNewValues() {
        preference.newValues.clear();
        int n = adapter.getCount();
        SparseBooleanArray checkedPositions = listView.getCheckedItemPositions();
        for (int i = 0; i < n; i++) {
            if (checkedPositions.get(i) == true) {
                preference.newValues.add((String) preference.entryValues[Arrays.asList(preference.entries).indexOf(adapter.getItem(i))]);
            }
        }
    }


    @Override
    protected void onPrepareDialogBuilder(androidx.appcompat.app.AlertDialog.Builder builder) {
        super.onPrepareDialogBuilder(builder);

        CharSequence[] mEntries = preference.getEntries();
        CharSequence[] mEntryValues = preference.entryValues;

        if (mEntries == null || mEntryValues == null) {
            throw new IllegalStateException("MultiSelectListPreference requires an entries array and " + "an entryValues array.");
        }

        String[] entries = new String[mEntries.length];
        for (int i = 0; i < mEntries.length; i++) {
            entries[i] = mEntries[i].toString();
        }

        boolean[] selectedItems = getSelectedItems();

        ArrayList<String> orderedList = new ArrayList<>();
        int n = selectedItems.length;

        for (String value : preference.values) {
            int index = Arrays.asList(mEntryValues).indexOf(value);
            orderedList.add(mEntries[index].toString());
        }

        for (int i = 0; i < mEntries.length; i++) {
            if (!preference.values.contains(mEntryValues[i])) orderedList.add(mEntries[i].toString());
        }

        adapter = new ArrayAdapter<>(getContext(), R.layout.item_list_preference_multi_drag, R.id.text, orderedList);
        listView = new DragSortListView(getContext(), null);
        listView.setAdapter(adapter);

        listView.setDropListener(onDrop);
        listView.setDragEnabled(true);
        listView.setFloatAlpha(0.8f);

        DragSortController controller = new DragSortController(listView);
        controller.setDragHandleId(R.id.drag_handle);
        controller.setRemoveEnabled(false);
        controller.setSortEnabled(true);
        controller.setBackgroundColor(0xFFFFFF);
        controller.setDragInitMode(DragSortController.ON_DOWN);

        listView.setFloatViewManager(controller);
        listView.setOnTouchListener(controller);


        builder.setView(listView);
        listView.setOnItemClickListener(new AdapterView.OnItemClickListener() {

            @Override
            public void onItemClick(AdapterView<?> arg0, View arg1, int arg2, long arg3) {
                mPreferenceChanged = true;
                refreshNewValues();
            }
        });

        listView.setChoiceMode(ListView.CHOICE_MODE_MULTIPLE);
        for (int i = 0; i < n; i++) {
            listView.setItemChecked(i, i < preference.values.size());
        }

        /*
         * boolean [] checkedItems = getSelectedItems();
         * builder.setMultiChoiceItems(mEntries, checkedItems,
         * new DialogInterface.OnMultiChoiceClickListener() {
         * public void onClick(DialogInterface dialog, int which, boolean
         * isChecked) {
         * if (isChecked) {
         * mPreferenceChanged |= mNewValues.add(mEntryValues[which].toString());
         * } else {
         * mPreferenceChanged |=
         * mNewValues.remove(mEntryValues[which].toString());
         * }
         * }
         * });
         */
        preference.newValues.clear();
        preference.newValues.addAll(preference.values);
    }

        @Override
        public void onDialogClosed ( boolean positiveResult)
        {
            if (positiveResult && mPreferenceChanged) {
                refreshNewValues();
                final List<String> values = preference.newValues;
                if (preference.callChangeListener(values) && values.size() > 0) {
                    preference.setValues(values);
                }
            }
            mPreferenceChanged = false;
        }
}

File: app/src/main/java/de/baumann/browser/view/sortlistpreference/MultiSelectDragListPreferenceDialog.kt
package de.baumann.browser.view.sortlistpreference

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.AdapterView.OnItemClickListener
import android.widget.ArrayAdapter
import android.widget.ImageView
import android.widget.ListView
import android.widget.TextView
import androidx.appcompat.app.AlertDialog
import androidx.preference.PreferenceDialogFragmentCompat
import de.baumann.browser.Ninja.R
import de.baumann.browser.view.sortlistpreference.DragSortListView.DropListener
import java.util.*

class MultiSelectDragListPreferenceDialog(
        private val preference: MultiSelectDragListPreference,
        private val shouldFinishActivityAfterDismissed: Boolean
) : PreferenceDialogFragmentCompat() {
    private lateinit var listView: DragSortListView
    private val selectedItems: BooleanArray
        private get() {
            val entries = preference.entryValues
            val entryCount = entries.size
            val values = preference.getValues()
            val result = BooleanArray(entryCount)
            for (i in 0 until entryCount) {
                result[i] = values.contains(entries[i].toString())
            }
            return result
        }
    private lateinit var iconListAdapter: ArrayAdapter<PreferenceItemInfo>
    private var mPreferenceChanged = false
    private val onDrop = DropListener { from, to ->
        if (from != to) {
            val item = iconListAdapter.getItem(from)
            iconListAdapter.remove(item)
            iconListAdapter.insert(item, to)
            listView.moveCheckState(from, to)
            mPreferenceChanged = true
            refreshNewValues()
        }
    }

    init {
        val b = Bundle()
        b.putString(ARG_KEY, preference.key)
        arguments = b
    }

    private fun refreshNewValues() {
        preference.newValues.clear()
        val n = iconListAdapter.count
        val checkedPositions = listView.checkedItemPositions
        for (i in 0 until n) {
            if (checkedPositions[i]) {
                preference.newValues.add(preference.entryValues[listOf(*preference.entries).indexOf(iconListAdapter.getItem(i))] as String)
            }
        }
    }

    override fun onPrepareDialogBuilder(builder: AlertDialog.Builder) {
        super.onPrepareDialogBuilder(builder)
        val mEntries = preference.getEntries()
        val mEntryValues = preference.entryValues
        val itemInfos = mEntries.mapIndexed { index, entry ->
            PreferenceItemInfo(
                    entry.toString(),
                    preference.entryIconIds[index]?.toString()?.toInt() ?: 0
            )
        }

        check(!(mEntries == null || mEntryValues == null)) { "MultiSelectListPreference requires an entries array and " + "an entryValues array." }

        val entries = arrayOfNulls<String>(mEntries.size)
        for (i in mEntries.indices) {
            entries[i] = mEntries[i].toString()
        }

        val selectedItems = selectedItems
        val orderedList = ArrayList<PreferenceItemInfo>()
        val n = selectedItems.size
        for (value in preference.values) {
            val index = listOf(*mEntryValues).indexOf(value)
            orderedList.add(itemInfos[index])
        }

        for (i in mEntries.indices) {
            if (!preference.values.contains(mEntryValues[i])) orderedList.add(itemInfos[i])
        }

        iconListAdapter = createItemAdapter(orderedList)
        listView = DragSortListView(context, null).apply {
            adapter = iconListAdapter
            setDropListener(onDrop)
            isDragEnabled = true
            floatAlpha = 0.8f
            choiceMode = ListView.CHOICE_MODE_MULTIPLE

            for (i in 0 until n) {
                setItemChecked(i, i < preference.values.size)
            }

            onItemClickListener = OnItemClickListener { _, _, _, _ ->
                mPreferenceChanged = true
                refreshNewValues()
            }

            builder.setView(this)
        }

        val controller = DragSortController(listView).apply {
            setDragHandleId(R.id.drag_handle)
            isRemoveEnabled = false
            isSortEnabled = true
            setBackgroundColor(0xFFFFFF)
            dragInitMode = DragSortController.ON_DOWN
        }

        listView.setFloatViewManager(controller)
        listView.setOnTouchListener(controller)


        preference.newValues.clear()
        preference.newValues.addAll(preference.values)
    }

    private fun createItemAdapter(itemList: List<PreferenceItemInfo>): ArrayAdapter<PreferenceItemInfo> {
        return object: ArrayAdapter<PreferenceItemInfo>(requireContext(), R.layout.item_list_preference_multi_drag, R.id.text, itemList) {
            override fun getView(position: Int, convertView: View?, parent: ViewGroup): View {
                val item = getItem(position)
                val view = convertView ?: LayoutInflater.from(context).inflate(R.layout.item_list_preference_multi_drag, parent, false) as CheckableMultiSelectLayout
                val imageView = view.findViewById<ImageView>(R.id.icon)
                val iconResId = item?.iconResId ?: 0
                if (iconResId != 0) {
                    imageView.setImageResource(iconResId)
                }
                view.findViewById<TextView>(R.id.text).text = item?.title ?: ""
                return view
            }
        }
    }

    override fun onDialogClosed(positiveResult: Boolean) {
        if (positiveResult && mPreferenceChanged) {
            refreshNewValues()
            val values = preference.newValues
            if (preference.callChangeListener(values) && values.size > 0) {
                preference.setValues(values)
            }
        }
        mPreferenceChanged = false

        if (shouldFinishActivityAfterDismissed) {
            activity?.finish()
        }
    }
}

private class PreferenceItemInfo(val title: String, val iconResId: Int)